{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# variant calling tutorials: from large vcf to de novo mutations\n",
    "\n",
    "this tutorial is under /home/hq2130/tutorials/vcf\n",
    "\n",
    "## We need to describe mappability filter and anything specific to WGS analysis. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# split large vcf into vcf trios with only de novo variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import gzip\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "vcf_name = 'cmbgvcf.rawvariants.recalibrated.vcf.hg19_multianno.vcf.gz'\n",
    "ped = '../src/CHD_MedExomeKit.ped'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def check_denovo(trio):\n",
    "    proband= trio[0].split(':')\n",
    "    Father = trio[1].split(':')\n",
    "    Mother = trio[2].split(':')\n",
    "    proband_GT = proband[0]\n",
    "    Father_GT = Father[0]\n",
    "    Mother_GT = Mother[0]\n",
    "    if proband_GT in {'./.','0/0'}: #proband genotype missing or not denovo\n",
    "        return False\n",
    "    else:#GT:AD:DP:GQ:PL    0/1:16,9:.:99:205,0,501\n",
    "        if Father_GT == '0/0' and Mother_GT == '0/0':\n",
    "            return True\n",
    "        else:\n",
    "            return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1-04687': ['1-04687', '1-04687-01', '1-04687-02'],\n",
       " '1-07134': ['1-07134', '1-07134-01', '1-07134-02']}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "with gzip.open(vcf_name, 'rt') as f:\n",
    "    for line in f: \n",
    "        if line[:6]=='#CHROM': \n",
    "            samples = line.split()\n",
    "            break\n",
    "            \n",
    "vcfid2sample = {} # if we want to convert vcf id to different names such as SRR001 to 1-0001\n",
    "family2id = {}\n",
    "with open('../src/CHD_MedExomeKit.ped') as f:\n",
    "    head = f.readline()\n",
    "    for line in f:\n",
    "        family, proband, father, mother = line.strip().split('\\t')[:4]\n",
    "        if proband in samples and father in samples and mother in samples:\n",
    "            \n",
    "            family2id[family] = [proband, father, mother]\n",
    "family2id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosome 1\n",
      "Chromosome 2\n",
      "Chromosome 3\n",
      "Chromosome 4\n",
      "Chromosome 5\n",
      "Chromosome 6\n",
      "Chromosome 7\n",
      "Chromosome 8\n",
      "Chromosome 9\n",
      "Chromosome 10\n",
      "Chromosome 11\n",
      "Chromosome 12\n",
      "Chromosome 13\n",
      "Chromosome 14\n",
      "Chromosome 15\n",
      "Chromosome 16\n",
      "Chromosome 17\n",
      "Chromosome 18\n",
      "Chromosome 19\n",
      "Chromosome 20\n",
      "Chromosome 21\n",
      "Chromosome 22\n",
      "Chromosome X\n",
      "Chromosome Y\n"
     ]
    }
   ],
   "source": [
    "with gzip.open(vcf_name, 'rt') as f:\n",
    "    head = []\n",
    "    ChrCount=0\n",
    "    for line in f:\n",
    "        if line.startswith('#'):\n",
    "            if line.startswith('#CHROM'):\n",
    "                vcf_samples = line.strip().split()\n",
    "                sample2idx = {}\n",
    "                for idx, sample in enumerate(vcf_samples):\n",
    "                    sample = vcfid2sample.get(sample, sample)\n",
    "                    sample2idx[sample] = idx\n",
    "                vcf_files = []\n",
    "                for proband in family2id:\n",
    "                    proband, father, mother = family2id[proband]\n",
    "                    if proband in sample2idx and father in sample2idx and mother in sample2idx:\n",
    "                        filename = 'vcf_denovo/' + '_'.join([proband, father, mother]) + '.vcf'\n",
    "                        vcf_files.append(open(filename, 'wt'))\n",
    "                n = len(vcf_files)\n",
    "                for i in range(n):\n",
    "                    proband, father, mother = vcf_files[i].name.split('/')[-1].split('.vcf')[0].split('_')\n",
    "                    vcf_files[i].write(''.join(head))\n",
    "                    vcf_files[i].write('\\t'.join(line.strip().split('\\t')[:9] + [proband, father, mother])+'\\n')  \n",
    "            else:\n",
    "                head.append(line)\n",
    "        else:\n",
    "            data = line.strip().split('\\t')\n",
    "            # verbose  \n",
    "            ChrPresent = data[0]\n",
    "            if ChrCount != ChrPresent:\n",
    "                print(\"Chromosome \"+str(ChrPresent))\n",
    "                ChrCount=ChrPresent\n",
    "                \n",
    "            for i in range(n):\n",
    "                proband, father, mother = vcf_files[i].name.split('/')[-1].split('.vcf')[0].split('_')\n",
    "                trio = [data[sample2idx[proband]], data[sample2idx[father]], data[sample2idx[mother]]]\n",
    "                if check_denovo(trio):\n",
    "            \n",
    "                    # filter AC larger than 60 \n",
    "                    INFOcolumnList=data[7].split(\";\")\n",
    "                    INFOdict={}\n",
    "                    for element in INFOcolumnList:\n",
    "                        if '=' in element:\n",
    "                            FieldName,FieldValue=element.split('=',1)\n",
    "                            INFOdict[FieldName]=FieldValue\n",
    "                            if FieldName == 'AC':\n",
    "                                break\n",
    "                    AC = sum(map(int, INFOdict['AC'].split(',')))\n",
    "                    if AC >= 60:\n",
    "                        continue\n",
    "                \n",
    "            \n",
    "                    vcf_files[i].write('\\t'.join(data[:9]+trio)+'\\n')\n",
    "            \n",
    "for fh in vcf_files:\n",
    "    fh.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# aggregate denovo trio vcfs into one csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "import csv\n",
    "\n",
    "def info_list(INFOstring, head):    \n",
    "    INFOcolumnList=INFOstring.split(\";\")\n",
    "    INFOdict={}\n",
    "    for element in INFOcolumnList:\n",
    "        if '=' in element:\n",
    "            FieldName,FieldValue=element.split('=',1)\n",
    "            INFOdict[FieldName]=FieldValue           \n",
    "    result = []\n",
    "    for e in head:\n",
    "        result.append(INFOdict.get(e,'NA'))\n",
    "    return result\n",
    "\n",
    "\n",
    "\n",
    "denovo_dir = 'vcf_denovo/'\n",
    "\n",
    "head = set()\n",
    "mutation = {}\n",
    "mutation_counts = defaultdict(lambda: [0, 0])\n",
    "for e in os.listdir(denovo_dir):\n",
    "    if e.endswith('.vcf'):\n",
    "        with open(denovo_dir +'/'+ e) as f1:\n",
    "            proband, father_ID, mother_ID= e.rsplit('.vcf')[0].split('_')\n",
    "            \n",
    "            for line in f1:\n",
    "                if line[0] =='#':\n",
    "                    if line.startswith('##INFO=<ID='):\n",
    "                        head.add(line.split('=')[2].split(',')[0])\n",
    "                else:\n",
    "                    info = line.strip().split()\n",
    "                    chrom = info[0]\n",
    "                    pos = info[1]\n",
    "                    INFOstring = info[7]\n",
    "                    format = info[8]\n",
    "\n",
    "                    variant_id = '_'.join([proband, chrom, pos])\n",
    "                    \n",
    "\n",
    "                    proband_geno = info[9].split(':')[0]\n",
    "                    proband_info = info[9]\n",
    "                    Father_geno = info[10].split(':')[0]\n",
    "                    Father_info = info[10]\n",
    "                    Mother_geno = info[11].split(':')[0]\n",
    "                    Mother_info = info[11]\n",
    "                    parents_list = father_ID + '(' + Father_info + ')' + ',' + mother_ID +'('+ Mother_info +')'\n",
    "                    \n",
    "                    variant_id_2 = '_'.join([chrom, pos])\n",
    "                    mutation_counts[variant_id_2][0] += 1\n",
    "\n",
    "                    if variant_id in mutation:\n",
    "                        print('same varaint in same proband again:' , variant_id)\n",
    "                    else:\n",
    "                        mutation[variant_id] = [format, proband+'('+ proband_info +')', parents_list] + info[:8]\n",
    "head = list(head)\n",
    "head.sort()\n",
    "with open('csv_denovo/test0713.csv','wt') as f:\n",
    "    w = csv.writer(f)\n",
    "    w.writerow(['CHROM','POS', 'ID', 'REF', 'ALT', 'QUAL','FILTER','FORMAT', \n",
    "                'proband', 'parents', 'het', 'homo'] + head)\n",
    "    for variant_id, value in mutation.items():\n",
    "        variant_id_2 = '_'.join(variant_id.split('_')[-2:])\n",
    "        content =  value[3:-1] + value[:3] + mutation_counts[variant_id_2] + info_list(value[-1], head)\n",
    "        w.writerow(content)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# filter for high confident denovo mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import datetime\n",
    "import yaml\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "exon True\n",
      "exon_flag ['splicing', 'exonic', 'exonic,splicing']\n",
      "max_cohort_AC 6\n",
      "max_AC_total 6\n",
      "max_population_AF {'ESP': 0.001, '1KG': 0.001, 'ExAC': 0.001}\n",
      "excluded_gene ['MUC', 'HLA']\n",
      "excluded_chrom ['GL']\n",
      "max_Multiallelic 3\n",
      "max_seqdup 0.99\n",
      "min_proband_AD 5\n",
      "min_proband_PL 60\n",
      "min_proband_alt_freq_tier1 0.2\n",
      "min_proband_alt_freq_tier2 0.2\n",
      "min_parents_DP 10\n",
      "min_parents_GQ 30\n",
      "max_parents_alt_freq 0.05\n",
      "max_FS 25\n",
      "min_QD 2\n",
      "max_FS 25\n",
      "min_QD 1\n",
      "min_ReadPosRankSum -3\n"
     ]
    }
   ],
   "source": [
    "now = datetime.datetime.now().strftime(\"%Y-%m-%d\")\n",
    "with open('../src/denovo_coding_soft.yml', 'r') as ymlfile:\n",
    "    cfg = yaml.load(ymlfile)\n",
    "\n",
    "csv_name = 'csv_denovo/test0713.csv'\n",
    "\n",
    "with open(csv_name.rsplit('/', 1)[0] + '/filter_parameter_' + now +'.txt', 'wt') as fw:\n",
    "    for section in cfg:\n",
    "        for key, value in cfg[section].items():\n",
    "            print(key, value)\n",
    "            fw.write('    %s %s: %s\\n' % (type(value), key, value))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# VarFunc exonic\n",
    "\n",
    "def info_filter(variants, snps, parameter):\n",
    "    '''\n",
    "    functions that flag the variant of FS, QD\n",
    "    '''\n",
    "    p = True\n",
    "\n",
    "    # freq filter\n",
    "    ExACfreq = max(map(float,[0] + [rate for rate in variants.get('ExAC_ALL','0').split(',') if rate not in {'NA','.'} ]))\n",
    "    \n",
    "    if ExACfreq > parameter['gene_filter']['max_population_AF']['ExAC']:\n",
    "        p = False\n",
    "\n",
    "    # exon filter\n",
    "    if parameter['gene_filter']['exon']:\n",
    "        if variants['Func.refGene'] not in parameter['gene_filter']['exon_flag']:\n",
    "            p = False\n",
    "\n",
    "    if variants['QD'] == 'NA':\n",
    "        variants['QD'] = 100\n",
    "    if snps:\n",
    "        if float(variants['FS']) > parameter['snps']['max_FS']:\n",
    "            p = False\n",
    "        if variants['QD'] == 'NA' or float(variants['QD']) < parameter['snps']['min_QD']:\n",
    "            p = False \n",
    "    else:\n",
    "        if float(variants['FS']) > parameter['indels']['max_FS']:\n",
    "            p = False\n",
    "        if float(variants['QD']) < parameter['indels']['min_QD']:\n",
    "            p = False   \n",
    "        if variants['ReadPosRankSum'] != 'NA' and float(variants['ReadPosRankSum']) < parameter['indels']['min_ReadPosRankSum']:\n",
    "            p = False   \n",
    "\n",
    "    return p\n",
    "\n",
    "def variants_filter(filename, parameter):\n",
    "    with open(filename,'rt') as denovo_f:\n",
    "        denovo_r = csv.reader(denovo_f)\n",
    "        head = next(denovo_r)\n",
    "\n",
    "        denovo_snp, denovo_indel =[], []\n",
    "        \n",
    "        denovo_write= open(filename.rsplit('_', 1)[0] + '_filtered_'+now+'.csv','wt')\n",
    "        w= csv.writer(denovo_write)\n",
    "        w.writerow(head) \n",
    "\n",
    "        denovo2_write= open(filename.rsplit('_', 1)[0] + '_tie2.csv','wt')\n",
    "        w2 = csv.writer(denovo2_write)\n",
    "        w2.writerow(head) \n",
    "\n",
    "        \n",
    "        idx = dict(zip(head, range(len(head))))\n",
    "        for line in denovo_r:\n",
    "\n",
    "            variant_pass = True\n",
    "            variant = dict(zip(head, line))\n",
    "            chrom = variant['CHROM']\n",
    "            pos = variant['POS']\n",
    "            \n",
    "            \n",
    "            ref = variant['REF']\n",
    "            alt = variant['ALT']\n",
    "            format = variant['FORMAT']\n",
    "            probands = variant['proband'].split(';')\n",
    "            parents = variant['parents'].split(';')\n",
    "            AC = variant['AC'].split(',') \n",
    "            \n",
    "            # remove multi-allel sites\n",
    "            if len(alt.split(',')) > parameter['gene_filter']['max_Multiallelic'] : \n",
    "                variant_pass = False\n",
    "            #print('0', variant_pass)\n",
    "            # remove know problem genes\n",
    "            for e in parameter['gene_filter']['excluded_chrom']:\n",
    "                if e in chrom:\n",
    "                    variant_pass = False\n",
    "\n",
    "            for e in parameter['gene_filter']['excluded_gene']:\n",
    "                if variant['Gene.refGene'].startswith(e):\n",
    "                    variant_pass = False\n",
    "\n",
    "            # remove high segdup score  \n",
    "            segdup = 0 if variant.get('genomicSuperDups') in {'NA','.'} else float(variant.get('genomicSuperDups',':0').split(':')[1].split('-')[0])\n",
    "            if segdup > parameter['gene_filter']['max_seqdup']: \n",
    "                variant_pass = False\n",
    "            #print('a', variant_pass)\n",
    "            \n",
    "            for i in range(len(probands)) :\n",
    "                newline = list(line)\n",
    "                proband = probands[i]\n",
    "                newline[idx['proband']] = proband\n",
    "                parent = parents[i] \n",
    "                newline[idx['parents']] = parent\n",
    "                ID, geno = proband.split('(')\n",
    "                GT = geno[:-1].split(':')[0]\n",
    "                _, alt_idx = map(int,GT.split('/'))\n",
    "                newline[idx['ALT']] = alt.split(',')[alt_idx-1]   \n",
    "\n",
    "\n",
    "                real_AC = int(AC[alt_idx-1])\n",
    "                if real_AC > parameter['gene_filter']['max_cohort_AC']:\n",
    "                    variant_pass = False\n",
    "                total_AC = sum(map(int, AC))\n",
    "                if total_AC > 2 * parameter['gene_filter']['max_cohort_AC']:\n",
    "                    variant_pass = False\n",
    "\n",
    "                snp = len(newline[idx['ALT']]) == len(ref) and newline[idx['ALT']] != '*'\n",
    "\n",
    "\n",
    "                \n",
    "                if not info_filter(variant, snp, parameter):\n",
    "                    variant_pass = False\n",
    "                \n",
    "                #print('b', variant_pass)\n",
    "                variant_id = '_'.join([ID,chrom,pos])\n",
    "\n",
    "                # change varclass format\n",
    "                varclass = variant['ExonicFunc.refGene']\n",
    "                if len(variant['ExonicFunc.refGene'].split(',')) > 1:\n",
    "                    varclass = variant['ExonicFunc.refGene'].split(',')[alt_idx-1]\n",
    "                    newline[idx['ExonicFunc.refGene']] = varclass\n",
    "               \n",
    "                #print('c', variant_pass, reads_filter(format, proband, parent, parameter))\n",
    "                if snp: \n",
    "                    if variant_pass and reads_filter(format, proband, parent, parameter):\n",
    "                        denovo_snp.append(variant_id)\n",
    "                        w.writerow(newline)  \n",
    "                    elif variant_pass:# and check_damage(variant, varclass):\n",
    "                        w2.writerow(newline)  \n",
    "\n",
    "                else:\n",
    "                    if variant_pass and reads_filter(format, proband, parent, parameter):\n",
    "                        denovo_indel.append( variant_id)\n",
    "                        w.writerow(newline)   \n",
    "                    elif variant_pass:# and check_damage(variant, varclass):\n",
    "                        w2.writerow(newline)              \n",
    "\n",
    "        denovo_write.close()\n",
    "        denovo2_write.close()\n",
    "        denovo_f.close()\n",
    "    return denovo_snp, denovo_indel\n",
    "\n",
    "def check_damage(variants, varclass):\n",
    "    '''\n",
    "    function return if a variant if damage \n",
    "    '''\n",
    "    damage = False\n",
    "    damage_key = {'frameshift','stop','splicing'}\n",
    "    for key in damage_key:\n",
    "        if varclass.startswith(key): \n",
    "            damage = True\n",
    "    \n",
    "    if 'nonsynonymous_SNV' in  varclass:\n",
    "        if 'D' in variants['MetaSVM_pred']:\n",
    "            damage = True\n",
    "        elif 'D' in variants['Polyphen2_HDIV_pred']:\n",
    "            cadd = max(map(float,[ rate  for rate in variants['CADD_phred'].split(',') if rate != '.' ]))\n",
    "            if cadd >= 15:\n",
    "                damage = True\n",
    "    return damage\n",
    "\n",
    "\n",
    "def get_dp(dp, ad):\n",
    "    '''reformat dp\n",
    "    '''\n",
    "    if dp == '.':\n",
    "        dp = sum(map(int, ad.split(',')))\n",
    "    dp = max([1, float(dp)])\n",
    "    return dp\n",
    "\n",
    "\n",
    "\n",
    "def reads_filter(format, proband, parents, parameter): # R,A used for checking indel AC is [1,1]\n",
    "    reads_pass = True\n",
    "    ID, geno =  proband.split('(')\n",
    "    father,mother = parents.split('),')\n",
    "    fatherID,fathergeno = father.split('(')\n",
    "    motherID,mothergeno = mother[:-1].split('(')\n",
    "\n",
    "    #print parents\n",
    "    proband_format = dict(zip(format.split(':'), geno[:-1].split(':')))\n",
    "    GT,AD,DP,GQ,PL =  proband_format['GT'], proband_format['AD'], proband_format['DP'], proband_format['GQ'], proband_format['PL']\n",
    "    father_format = dict(zip(format.split(':'), fathergeno.split(':')))\n",
    "    fGT,fAD,fDP,fGQ,fPL =  father_format['GT'], father_format['AD'], father_format['DP'], father_format['GQ'], father_format['PL']\n",
    "    mother_format = dict(zip(format.split(':'), mothergeno[:-1].split(':')))\n",
    "    mGT,mAD,mDP,mGQ,mPL  =  mother_format['GT'], mother_format['AD'], mother_format['DP'], mother_format['GQ'], mother_format['PL']\n",
    "\n",
    "        \n",
    "    ref_idx, alt_idx = map(int,GT.split('/'))\n",
    "    \n",
    "    # parents filter\n",
    "    if  mDP == '.':\n",
    "        mDP = max([1, sum(map(int, mAD.split(',')))])\n",
    "    if  fDP == '.':\n",
    "        fDP = max([1, sum(map(int, fAD.split(',')))])\n",
    "    if  DP in {'.','0'}:\n",
    "        if DP == '.':\n",
    "            DP = sum(map(int, AD.split(',')))\n",
    "        else:\n",
    "            print (proband, DP, 'proband dp . ??')\n",
    "            DP = 1  \n",
    "                \n",
    "    if float(mDP) < 1:\n",
    "        mDP = 1\n",
    "    if float(fDP) < 1:\n",
    "        fDP = 1\n",
    "    Fmaf = int(fAD.split(',')[alt_idx])/float(fDP)\n",
    "    Mmaf = int(mAD.split(',')[alt_idx])/float(mDP) \n",
    "\n",
    "     # parents filter\n",
    "    #print('xa', reads_pass, max(Fmaf, Mmaf), parameter['reads_filter']['max_parents_alt_freq'])\n",
    "    if max(Fmaf, Mmaf) > parameter['reads_filter']['max_parents_alt_freq']:\n",
    "        reads_pass = False\n",
    "    if min(int(mDP), int(fDP)) < parameter['reads_filter']['min_parents_DP']:\n",
    "        reads_pass = False\n",
    "    if min(int(mGQ), int(fGQ)) < parameter['reads_filter']['min_parents_GQ']:\n",
    "        reads_pass = False\n",
    "    # prband filter\n",
    "    proband_AD = list(map(int,AD.split(',')))\n",
    "    proband_DP = max([1, float(DP)])\n",
    "    proband_PL = int(PL.split(',')[0])\n",
    "    #print proband_DP, proband\n",
    "    proband_maf = proband_AD[alt_idx]/proband_DP\n",
    "    \n",
    "   # print('x', reads_pass)\n",
    "\n",
    "    if proband_PL < parameter['reads_filter']['min_proband_PL']: #require proband_PL of 0/0 larger than p, which is probably of 0/0 is 10^-p\n",
    "        reads_pass = False\n",
    "    \n",
    "    #print('x', reads_pass, proband_AD[alt_idx], parameter['reads_filter']['min_proband_AD'])\n",
    "    \n",
    "    if proband_AD[alt_idx] <  parameter['reads_filter']['min_proband_AD']:\n",
    "        reads_pass = False\n",
    "    \n",
    "    if proband_AD[alt_idx] < 10:\n",
    "        if proband_maf < parameter['reads_filter']['min_proband_alt_freq_tier1']:\n",
    "            reads_pass = False\n",
    "    else:\n",
    "        if proband_maf < parameter['reads_filter']['min_proband_alt_freq_tier2']:\n",
    "            reads_pass = False\n",
    "    return  reads_pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2 denovo snp, 1 denovo indels\n"
     ]
    }
   ],
   "source": [
    "denovo_snp, denovo_indel = variants_filter(csv_name, cfg)\n",
    "print ('%s denovo snp, %s denovo indels' % (len(denovo_snp), len(denovo_indel)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
